### PR TITLE
Improve MCM conformance in comm=ofi.

### PR DESCRIPTION
**Pull request by gbtitus**
This is a collection of changes that improve MCM conformance in the
libfabric-based comm layer.  The major changes are to favor delivery-
complete completion semantics as our solution for MCM conformance over
message ordering settings, and to improve how we ensure non-fetching
AMOs done by means of nonblocking AMs are complete as required by the
MCM.

The original "solution" to the problem of ensuring AMO completion was
added in PR #15712 and involved using non-blocking AMs, tracking the
number of such outstanding AMs and their target nodes, and following up
when needed with blocking no-op AMs to ensure the non-blocking ones were
done.  Unfortunately this was both heavyweight because it involved
allocating and maintaining a numNodes-sized bitmap, and broken because
it only forced order among AMOs targeting the same node, not all nodes
as the MCM requires.  So here, we remove this and instead solve the
problem by using blocking AMs so that we know when the target memory has
been updated, but not waiting for the 'done' indicators resulting from
those until we do the next thing with MCM implications.

Relatedly, this also adds `chpl_comm_task_create()` as a sibling of the
task fence function `chpl_comm_task_end()`, and calls this new function
from `_upEndCount()` before we create child tasks or groups of same.  We
only need the parent to call this because its purpose is to fence the
parent's "recent" remote operations (including non-fetching AMOs) so
that their results are definitely visible to children created later.
Because for now only comm=ofi needs this function, it's declared by
means of the usual mechanism for optional runtime functions and only ofi
defines it.

While here I also made some performance-related changes.  We now use the
"domain" threading model rather than the default "safe" one, which
reduces lock usage inside libfabric and its providers.  We also now
inject sending operations when possible instead of initiating them
normally, which speeds things up slightly.

This improves performance in some tests and reduces it in others, though
overall the effect seems positive.  (Some of the regressions seem to be
an unfortunate side effect of removing the bug related to non-fetching
AMO completion.)  The new create-task fence function may also allow us
to improve performance for regular PUTs in the future.

This resolves https://github.com/Cray/chapel-private/issues/1307.

### CODE REVIEW

File: modules/internal/ChapelBase.chpl [hunk 1]
```diff
@@ -1215,6 +1215,9 @@ module ChapelBase {
     delete _to_unmanaged(e);
   }
 
+  pragma "no doc"
```
Comment (ronawho) [inline modules/internal/ChapelBase.chpl:1218]: This module isn't documented, so "no doc" isn't needed. Can we put `chpl_comm_task_create` with `chpl_comm_task_end`?

Comment (ronawho) [reply modules/internal/ChapelBase.chpl:1218]: Can we put a doc on on both task_create and task_end saying something like:


create -- "called by the parent task before child tasks are spawned" 
end -- "called by the child task while ending"

at first it wasn't obvious to me that task_create was supposed to be called by the parent or the child I wonder if chpl_comm_task_pre_create is slightly more obvious in that regard, but i don't think it's worth changing

File: modules/internal/ChapelBase.chpl [hunk 2]
```diff
@@ -1237,6 +1240,9 @@ module ChapelBase {
     }
     if countRunningTasks {
       here.runningTaskCntAdd(1);  // decrement is in _waitEndCount()
+      chpl_comm_task_create();    // countRunningTasks is a proxy for "is local"
+                                  // here.  Comm layers are responsible for the
+                                  // remote case themselves.
```
Comment (ronawho) [inline modules/internal/ChapelBase.chpl:1245]: Just confirming -- this means that the comm layer will fence before doing an on-stmt, right?

File: modules/internal/ChapelBase.chpl (hunk 1 above)
Comment (ronawho) [reply modules/internal/ChapelBase.chpl:1218]: Ah, I see these are documented in the runtime already, maybe that's enough

**Review by ronawho — APPROVED** [review]

File: modules/internal/ChapelBase.chpl (hunk 1 above)
Comment (gbtitus) [reply modules/internal/ChapelBase.chpl:1218]: Ah, I think I actually stole a decl from `MemConsistency.chpl` to form that one, is why the "no doc" is there.  I'll remove it.

The decl and use of `chpl_comm_task_end()` are below this point in ChapelBase, so we'd actually want to move `task_end` to be with `task_create` in order to put them together.  But yeah, that would probably be good.

Comment (ronawho) [reply modules/internal/ChapelBase.chpl:1218]: I don't think chapel requires functions to be declared before they're used, but up to you where you want to place the externs.

File: modules/internal/ChapelBase.chpl (hunk 2 above)
Comment (gbtitus) [reply modules/internal/ChapelBase.chpl:1245]: Yes.

**Review by ronawho — APPROVED** [review]

**Pull request merged**